poc/vdaf_poplar1.py

```diff
@@ -151,16 +151,39 @@ def shard(Poplar1, measurement, nonce, rand):
         input_shares = list(zip(keys, corr_seed, corr_inner, corr_leaf))
         return (public_share, input_shares)
 
-    def is_valid(agg_param, previous_agg_params):
+    @classmethod
+    def get_ancestor(Poplar1, input, this_level, last_level):
         """
-        Checks if the level of `agg_param` appears anywhere in
-        `previous_agg_params`. Returns `False` if it does, and `True` otherwise.
+        Helper function to determine the prefix of `input` at `last_level`.
         """
-        (level, _) = agg_param
-        return all(
-            level != other_level
-            for (other_level, _) in previous_agg_params
-        )
+        return input >> (this_level - last_level)
+
+    @classmethod
+    def is_valid(Poplar1, agg_param, previous_agg_params):
+        """
+        Checks that levels are increasing between calls, and also enforces that
+        the prefixes at each level are suffixes of the previous level's
+        prefixes.
+        """
+        if len(previous_agg_params) < 1:
+            return True
+
+        (level, prefixes) = agg_param
+        (last_level, last_prefixes) = previous_agg_params[-1]
+        # The empty prefix 0 is always there.
+        last_prefixes_set = set(list(last_prefixes) + [0])
```

Comment:
See the related suggestion on the specification.
```suggestion
        last_prefixes_set = set(last_prefixes)
```

Reply:
Bump